File: lib/motel/movement_strategies/rotate.rb
# The Rotation MovementStrategy
#
# Copyright (C) 2010-2014 Mohammed Morsi <mo@morsi.org>
# Licensed under the AGPLv3 http://www.gnu.org/licenses/agpl.txt

require 'rjr/common'
require 'motel/movement_strategy'
require 'motel/mixins/movement_strategy/rotatable'

module Motel
module MovementStrategies
# Rotates a location around its own access at a specified speed.
class Rotate < MovementStrategy
  include Rotatable

  def initialize(args = {})
    init_rotation(args)
    super(args)
  end

  # Return boolean indicating if this movement strategy is valid
  def valid?
    valid_rotation?
  end

  # Return true if we should change ms due to rotation
  def change?(loc)
    change_due_to_rotation?(loc)
  end

  # Implementation of {Motel::MovementStrategy#move}
  def move(loc, elapsed_seconds)
    unless valid?
      ::RJR::Logger.warn "rotate movement strategy (#{rot_to_s}) not valid, "\
                         "not proceeding with move"
      return
    end

    ::RJR::Logger.debug "moving location #{loc.id} "\
                        "via rotate movement strategy #{rot_to_s}"

    rotate(loc, elapsed_seconds)
  end

  # Convert movement strategy to json representation and return it
  def to_json(*a)
    { 'json_class' => self.class.name,
      'data'       => { :step_delay => step_delay}.merge(rotation_json)
    }.to_json(*a)
  end

  # Convert movement strategy to human readable string and return it
  def to_s
    "rotate-(#{rot_to_s})"
  end

end # class Rotate
end # module MovementStrategies
end # module Motel

File: site/source/javascripts/omega/ship/movement/rotate.js
/* Omega JS Ship Rotation Movement Mixin
 *
 * Copyright (C) 2014 Mohammed Morsi <mo@morsi.org>
 *  Licensed under the AGPLv3 http://www.gnu.org/licenses/agpl.txt
 */

Omega.ShipRotationMovement = {
  _rotate : function(elapsed, invert){
    var rot_theta = invert ? (this.location.movement_strategy.rot_theta * -1) :
                              this.location.movement_strategy.rot_theta
    if(!rot_theta) return;

    var stop_angle = this.location.movement_strategy.stop_angle;
    if(this.location.angle_rotated + rot_theta > stop_angle)
      rot_theta = stop_angle - this.location.angle_rotated;

    var dist = rot_theta * elapsed / 1000;
    this.location.rotate_orientation(dist);
    this.location.angle_rotated += rot_theta;
  },

  _run_rotation_movement : function(page, elapsed, invert){
    var now     = new Date();
        elapsed = elapsed || (now - this.last_moved);

    this._rotate(elapsed, invert);

    this.update_gfx();
    this.last_moved = now;
    this.dispatchEvent({type : 'movement', data : this});
  }
};
